File: OOP_Inheritance.cpp
#include <bits/stdc++.h>

using namespace std;
class Person
{
private:
    string name;
    string address;

public:
    string getName()
    {
        return name;
    }
    string getAddress()
    {
        return address;
    }
    void setName(string name)
    {
        this->name = name;
    }
    void setAddress(string address)
    {
        this->address = address;
    }
    void in()
    {
        cout << name << " " << address << " ";
    }

    Person(string name, string address)
    {
        this->name = name;
        this->address = address;
    }
};

class Student : public Person
{
private:
    double gpa;

public:
    double getGPA()
    {
        return gpa;
    }
    void setGPA(double gpa)
    {
        this->gpa = gpa;
    }
    void in()
    { // function overriding
        Person::in();
        cout << gpa << fixed << setprecision(2) << endl;
    }
    Student(string name, string address, double gpa) : Person(name, address)
    {
        this->gpa = gpa;
    }
};

class A
{
public:
    A()
    {
        cout << "Constructor A \n!";
    }
    ~A()
    {
        cout << "Destructor A \n!";
    }
    void xinChao()
    {
        cout << "Hello A \n!";
    }
};
class B : public A
{
public:
    B()
    {
        cout << "Constructor B \n!";
    }
        ~B()
    {
        cout << "Destructor B \n!";
    }
    void xinChao()
    {
        cout << "Hello B \n!";
    }
};
class C : public B // multi level
{
public:
    C()
    {
        cout << "Constructor C \n!";
    }
        ~C()
    {
        cout << "Destructor C \n!";
    }
    void xinChao()
    {
        cout << "Hello C \n!";
    }
};
class D : public A,public B,public C {}; //multiple inheritance
int main()
{

    Student s("Tuanvv", "QuangBinh", 3.50);
    // s.setName("Tuanvv");
    // s.setAddress("QuangBinh");
    // s.setGPA(3.0);
    s.in();
    // s.Person::in();
    C obj; 
    obj.xinChao(); 

}

File: OOP_Inheritance.java
import java.util.*;

class Person {
    private String name, address;

    Person() {
    }

    Person(String name, String address) {
        this.name = name;
        this.address = address;
    }

    String getName() {
        return name;
    }

    String getAddress() {
        return address;
    }

    void setName(String name) {
        this.name = name;
    }

    void setAddress(String address) {
        this.address = address;
    }

    void in() {
        System.out.println(name + " " + address + " ");
    }
}

class Student extends Person {
    private double gpa;

    void setGPA(double gpa) {
        this.gpa = gpa;
    }

    double getGPA() {
        return gpa;
    }

    @Override
    void in() {
        super.in();
        System.out.println(gpa + " ");
    }

    Student(String name, String address, double gpa) {
        super(name, address);
        this.gpa = gpa;
    }

    Student() {
    }
}


// // Interface Person
// interface Person {
//     String getName();
//     String getAddress();
//     void setName(String name);
//     void setAddress(String address);
//     void in(); // in thông tin
// }

// // Interface GPAInfo (tạo thêm để mô phỏng đa kế thừa)
// interface GPAInfo {
//     void setGPA(double gpa);
//     double getGPA();
// }

// // Class Student implements 2 interface
// class Student implements Person, GPAInfo {
//     private String name;
//     private String address;
//     private double gpa;

//     // Constructors
//     Student() {}

//     Student(String name, String address, double gpa) {
//         this.name = name;
//         this.address = address;
//         this.gpa = gpa;
//     }

//     // Implement methods from Person
//     public String getName() {
//         return name;
//     }

//     public String getAddress() {
//         return address;
//     }

//     public void setName(String name) {
//         this.name = name;
//     }

//     public void setAddress(String address) {
//         this.address = address;
//     }

//     // Implement methods from GPAInfo
//     public void setGPA(double gpa) {
//         this.gpa = gpa;
//     }

//     public double getGPA() {
//         return gpa;
//     }

//     // Display method
//     public void in() {
//         System.out.println(name + " " + address);
//         System.out.printf("GPA: %.2f\n", gpa);
//     }
// }
public class OOP_Inheritance {
    public static void main(String args[]) {

        Student s = new Student();

        s.setName("Tuanvv");
        s.setAddress("QuangBinh");
        s.setGPA(3.0);
        s.in();

        Student ss = new Student("Datvv", "Dong Hoi", 3.6);
        ss.in();
    }
}
